platforms/ide/problems-api/src/main/java/org/gradle/internal/problems/failure/FailureFactory.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.problems.failure;
+
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+
+@ServiceScope(Scope.Global.class)
+public interface FailureFactory {
+
+    Failure create(Throwable failure);
+
+    Failure create(Throwable failure, StackTraceClassifier classifier);
```

Comment:
But we have the use case with the `calledFrom`, when stacktrace categorization is directly influenced by the caller. In the current implementation a part of the stacktrace is dropped up to the `calledFrom` frames.

https://github.com/gradle/gradle/blob/37ff26f9dd6be75af0cc1a97115fc5250862165d/platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/StackTraceSanitizer.java#L36-L48

Reply:
The caller can do this filtering itself after creating the `Failure`.

Reply:
I wanted to prevent callers of the failure factory from meddling with the stacktrace again. Otherwise, we risk that logic spreading back to callers again.

Reply:
I'm not suggesting the caller change the stack trace. I'm suggesting the caller use the information that is carried in the `Failure` to do the specific reporting they need to do.

Specifically, the deprecation handling stuff uses the stack trace to do two things:

1. Log the "location" where the deprecation happened: https://github.com/gradle/gradle/blob/5e1c4a80f797bcb23aea78f092be83ac744e593c/platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java#L121
2. Forward the stack to scans: https://github.com/gradle/gradle/blob/fe9636b2df332d5655097c1a087ebd7eeffac480/platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/DefaultDeprecatedUsageProgressDetails.java#L75

The first usage can be replaced with simply reporting the first "user" stack trace element.

The second usage can be replaced by forwarding only the "user" stack trace elements, at least until scans can take the categorized stack trace instead.

Reply:
I would split addressing this out into a follow up PR, to avoid getting bogged down in this issue. There's already a lot of value in this PR.